feat: upgrade Apache NiFi to v2.7.2

saidsef opened the pull request:
## What this PR does / why we need it:
Upgrades Apache NiFi to version 2.7.2 (stable). This update includes significant security enhancements, NiFi 2.x specific configuration adjustments, and a move towards Kubernetes-native features.

Key changes:
- **Upgraded NiFi to 2.7.2** (latest stable 2.x).
- **Kubernetes-native Leader Election:** Switched from ZooKeeper-based election to `KubernetesLeaderElectionManager` using Kubernetes Leases.
- **Secure by Default:** Enabled HTTPS (port 8443) for cluster communication, which is mandatory in NiFi 2.x.
- **Automated Security:** Fixed and enabled the `security.sh` script for automatic self-signed certificate generation on startup.
- **RBAC:** Added `ServiceAccount`, `Role`, and `RoleBinding` to support Lease management.
- **Namespace isolation:** Resources are now explicitly managed within the `nifi` namespace via Kustomize.
- **Cleanup:** Removed unused/deprecated configurations for Kafka and legacy Zookeeper clusters.

## Types of changes
- [✓] Bug fix (non-breaking change which fixes an issue)
- [✓] New feature (non-breaking change which adds functionality)
- [✓] Breaking change (NiFi 2.x upgrade and switch to HTTPS)

## How Has This Been Tested?
- Verified on `kind` (v1.30.0).
- Confirmed successful cluster initialization, leader election (via `leases`), and UI accessibility (HTTPS).

## Checklist:
- [✓] My change requires a change to the documentation.
- [✓] I have updated the documentation accordingly (README updated).
- [✓] I've read the CONTRIBUTION /CONTRIBUTING.md guide.
- [✓] All new and existing tests passed.

saidsef committed:
feat: upgrade Apache NiFi to v2.1.0

Full release notes [v2.1.0](https://cwiki.apache.org/confluence/display/NIFI/Release+Notes#ReleaseNotes-Version2.0.0)
Closes: #39

saidsef committed:
feat: refactor env vars and add proxy name

saidsef committed:
fix: trust store key type

fix: nifi config paths

saidsef committed:
feat: enable nifi secure script

fix: set nifi cluster set to true and readiness delay to 200

peter279k commented on deployment/nifi/configmap.yml:
```diff
@@ -3,12 +3,11 @@ kind: ConfigMap
   KEYSTORE_TYPE: "PKCS12"
   NIFI_ANALYTICS_PREDICT_ENABLED: "true"
-  NIFI_CLUSTER_IS_NODE: "true"
+  NIFI_CLUSTER_IS_NODE: "false"
```
This setting should be true and increase the `readinessProbe.initialDelaySeconds` setting in the `nifi/nifi.yml` file.

peter279k commented on deployment/nifi/nifi.yml:
```diff
@@ -137,7 +151,7 @@ spec:
           tcpSocket:
             port: cluster
-          initialDelaySeconds: 60
+          initialDelaySeconds: 90
```
After increasing the `readinessProbe.initialDelaySeconds`, it will be worked. And I configure `200` seconds.

peter279k replied:
The NiFi K8s cluster is completed now after doing above step:

![image](https://github.com/user-attachments/assets/61bd2c4e-5daa-4403-b2c8-50bfac713ba7)

saidsef commented:
Hi @peter279k , Thanks for looking into that.

I have off-line version with those values set - that's now pushed to the PR branch. However, I'm getting this waring/error on the second node:
```sh
Caused by: sun.security.validator.ValidatorException: PKIX path validation failed: java.security.cert.CertPathValidatorException: Path does not chain with any of the trust anchors
	at java.base/sun.security.validator.PKIXValidator.doValidate(PKIXValidator.java:318)
	at java.base/sun.security.validator.PKIXValidator.engineValidate(PKIXValidator.java:267)
	at java.base/sun.security.validator.Validator.validate(Validator.java:256)
	at java.base/sun.security.ssl.X509TrustManagerImpl.checkTrusted(X509TrustManagerImpl.java:230)
	at java.base/sun.security.ssl.X509TrustManagerImpl.checkServerTrusted(X509TrustManagerImpl.java:132)
	at org.apache.nifi.security.ssl.StandardX509ExtendedTrustManager.checkServerTrusted(StandardX509ExtendedTrustManager.java:61)
	at java.base/sun.security.ssl.CertificateMessage$T13CertificateConsumer.checkServerCerts(CertificateMessage.java:1302)
	... 24 common frames omitted
```

Do you see that your deployment?

peter279k commented:
> Hi @peter279k , Thanks for looking into that.
> 
> I have off-line version with those values set - that's now pushed to the PR branch. However, I'm getting this waring/error on the second node:
> 
> ```shell
> Caused by: sun.security.validator.ValidatorException: PKIX path validation failed: java.security.cert.CertPathValidatorException: Path does not chain with any of the trust anchors
> 	at java.base/sun.security.validator.PKIXValidator.doValidate(PKIXValidator.java:318)
> 	at java.base/sun.security.validator.PKIXValidator.engineValidate(PKIXValidator.java:267)
> 	at java.base/sun.security.validator.Validator.validate(Validator.java:256)
> 	at java.base/sun.security.ssl.X509TrustManagerImpl.checkTrusted(X509TrustManagerImpl.java:230)
> 	at java.base/sun.security.ssl.X509TrustManagerImpl.checkServerTrusted(X509TrustManagerImpl.java:132)
> 	at org.apache.nifi.security.ssl.StandardX509ExtendedTrustManager.checkServerTrusted(StandardX509ExtendedTrustManager.java:61)
> 	at java.base/sun.security.ssl.CertificateMessage$T13CertificateConsumer.checkServerCerts(CertificateMessage.java:1302)
> 	... 24 common frames omitted
> ```
> 
> Do you see that your deployment?

@saidsef, thanks for your reply and help me to resolve the issue.

I also have the same error that you mention and it seems that the problem is about the truststore configuration.

peter279k commented:
@saidsef, I also found this [reference](https://community.cloudera.com/t5/Support-Questions/NIFI-2-0-Cluster-Set-Up/td-p/388674) about ` PKIX path validation failed` issue when deploying the NiFi 2.0 cluster.

I also found that the first node and second node have different self-signed truststore certificate.

Could you do the same self-signed truststore cerificate in this NiFi cluster node?

peter279k commented:
> @saidsef, I also found this [reference](https://community.cloudera.com/t5/Support-Questions/NIFI-2-0-Cluster-Set-Up/td-p/388674) about ` PKIX path validation failed` issue when deploying the NiFi 2.0 cluster.
> 
> I also found that the first node and second node have different self-signed truststore certificate.
> 
> Could you use the same self-signed truststore cerificate in this NiFi cluster node?

@saidsef, I modified the `./nifi/hpa.yml` file and configure minReplicas and maxReplicas be `1`.

That is, it's only setup a single node NiFi cluster. And the `PKIX path validation failed` is not occurred.

Hexoplon commented:
@saidsef Have you tested moving to the k8s leader election implementation, instead of using zookeper? https://nifi.apache.org/nifi-docs/administration-guide.html#kubernetes-clustering

saidsef committed:
chore: update CuratorLeaderElectionManager config

saidsef committed:
Upgrade to Apache NiFi 2.7.2, enable secure clustering, and switch to Kubernetes leader election

saidsef committed:
fix(ci/kind): update k8s Kind Cluster configuration for consistency

saidsef committed:
Fix SSL cluster trust and SNI issues in NiFi 2.x using shared PVC and subPath mounts

saidsef committed:
Update NiFi configuration to allow any SNI and Proxy hosts by default

saidsef committed:
Force admin credentials initialization on startup

saidsef committed:
feat(k8s/pv): add pv for nifi certs

fix(nifi): set fsGroup in security context for StatefulSet

feat(nifi): add init container to fix permissions for certs volume

fix(nifi): adjust security context for certs and update mount paths

saidsef committed:
feat(nifi): upgrade to NiFi 2.7.2 with Kubernetes-native clustering

- Upgrade NiFi to 2.7.2 with secure HTTPS clustering enabled
- Replace ZooKeeper with KubernetesLeaderElectionManager and
  KubernetesConfigMapStateProvider for zero-dependency cluster state
- Add kind-config.yml: 2-node cluster with shared hostPath extraMounts
  enabling ReadWriteMany cert PVC across nodes
- Add configmap-statemgmt.yml: state-management.xml with WriteAhead local
  provider and KubernetesConfigMapStateProvider (no properties needed)
- Add configmap-authorizers.yml: single-user-authorizer (enforced by
  start.sh when NIFI_SINGLE_USER_CREDENTIALS_* env vars are set)
- Fix cert SAN: remove ip:${POD_IP}; wildcard dns:*.nifi.nifi.svc.cluster.local
  covers all pods without binding to ephemeral pod IPs
- Reduce probe initialDelaySeconds 600s→60s so both pods reach Ready state
  within the OrderedReady window
- Update PV/PVC to ReadWriteMany; add delete verb to RBAC for state cleanup
- Remove ZooKeeper deployments (zookeeper/ and zookeeper-cluster/ directories)

saidsef committed:
fix(ci): update kind cluster to v1.14.0 with kubectl and node image pinned

saidsef committed:
docs(readme): fix markdown style and remove stale ZooKeeper references
